presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/SchemaAccessControlRule.java

```diff
@@ -24,26 +29,85 @@
 public class SchemaAccessControlRule
 {
     private final boolean owner;
+    private final Optional<AccessMode> allow;
     private final Optional<Pattern> userRegex;
     private final Optional<Pattern> schemaRegex;
 
     @JsonCreator
     public SchemaAccessControlRule(
             @JsonProperty("owner") boolean owner,
+            @JsonProperty("allow") Optional<AccessMode> allow,
             @JsonProperty("user") Optional<Pattern> userRegex,
             @JsonProperty("schema") Optional<Pattern> schemaRegex)
     {
         this.owner = owner;
+        this.allow = requireNonNull(allow, "allow is null");
         this.userRegex = requireNonNull(userRegex, "userRegex is null");
         this.schemaRegex = requireNonNull(schemaRegex, "sourceRegex is null");
     }
 
-    public Optional<Boolean> match(String user, String schema)
+    public Optional<Boolean> isSchemaOwner(String user, String schema)
     {
         if (userRegex.map(regex -> regex.matcher(user).matches()).orElse(true) &&
                 schemaRegex.map(regex -> regex.matcher(schema).matches()).orElse(true)) {
             return Optional.of(owner);
         }
         return Optional.empty();
     }
+
+    public Optional<AccessMode> match(String user, String schema)
+    {
+        if (userRegex.map(regex -> regex.matcher(user).matches()).orElse(true) &&
+                schemaRegex.map(regex -> regex.matcher(schema).matches()).orElse(true)) {
+            // by default, if user did not defined visibility, assume schema is visible for back compatible
+            if (!allow.isPresent()) {
+                return Optional.of(AccessMode.VISIBLE);
+            }
+            return allow;
+        }
+        return Optional.empty();
+    }
+
+    public enum AccessMode
+    {
+        VISIBLE("visible"),
+        NONE("none");
+
+        private static final Map<String, AccessMode> modeByName = Maps.uniqueIndex(ImmutableList.copyOf(AccessMode.values()), AccessMode::toString);
+
+        private final String stringValue;
+
+        AccessMode(String stringValue)
+        {
+            this.stringValue = requireNonNull(stringValue, "stringValue is null");
+        }
+
+        @JsonValue
+        @Override
+        public String toString()
+        {
+            return stringValue;
+        }
+
+        @JsonCreator
+        public static AccessMode fromJson(Object value)
+        {
+            if (value instanceof String) {
+                AccessMode accessMode = modeByName.get(((String) value).toLowerCase(Locale.US));
+                if (accessMode != null) {
+                    return accessMode;
+                }
+            }
+
+            throw new IllegalArgumentException("Unknown " + AccessMode.class.getSimpleName() + ": " + value);
+        }
+
+        boolean implies(AccessMode other)
+        {
+            if (this == VISIBLE && other == NONE) {
```

Comment:
I would prefer to extract `AccessMode` from `CatalogAccessControlRule` that would have same value and same logic.

Reply:
so move ` CatalogAccessControlRule::AccessMode ` to ` security/AccessMode ` ?

Reply:
@kokosing ready for reveiw